tests/xdc/history_replication_signals_and_updates_test.go

```diff
@@ -785,33 +781,62 @@ func (task *hrsuTestExecutableTask) workflowId() string {
 
 // Update test utilities
 func (c *hrsuTestCluster) sendUpdateAndWaitUntilStage(ctx context.Context, updateId string, arg string, stage sdkclient.WorkflowUpdateStage) {
-	updateResponse := make(chan error)
-	processWorkflowTaskResponse := make(chan error)
+	updateErrCh := make(chan error)
 	go func() {
 		_, err := c.client.UpdateWorkflow(ctx, sdkclient.UpdateWorkflowOptions{
 			UpdateID:     updateId,
 			WorkflowID:   c.t.tv.WorkflowID(),
 			RunID:        c.t.tv.RunID(),
-			UpdateName:   "the-test-doesn't-use-this",
-			Args:         []interface{}{arg},
+			UpdateName:   c.t.tv.Any().String(),
+			Args:         []any{arg},
 			WaitForStage: stage,
 		})
-		c.t.s.NoError(err)
-		updateResponse <- err
+		updateErrCh <- err
 	}()
-	go func() {
-		// Blocks until the update request causes a WFT to be dispatched; then sends the update acceptance message
-		// required for the update request to return.
-		if stage == sdkclient.WorkflowUpdateStageCompleted {
-			processWorkflowTaskResponse <- c.pollAndAcceptCompleteUpdate(updateId)
-		} else if stage == sdkclient.WorkflowUpdateStageAccepted {
-			processWorkflowTaskResponse <- c.pollAndAcceptUpdate()
-		} else {
-			c.t.s.FailNow("invalid stage")
+
+	// Wait admitted to make sure that Update reached server, and added to registry.
+	// This guarantees following poll to get an Update request message.
+	c.waitUpdateAdmitted(c.t.tv, updateId)
+
+	// Blocks until the update request causes a WFT to be dispatched; then sends the update acceptance message
+	// required for the update request to return.
+	if stage == sdkclient.WorkflowUpdateStageCompleted {
+		err := c.pollAndAcceptCompleteUpdate(updateId)
+		c.t.s.NoError(err)
+	} else if stage == sdkclient.WorkflowUpdateStageAccepted {
+		err := c.pollAndAcceptUpdate()
+		c.t.s.NoError(err)
+	} else {
+		c.t.s.FailNow("invalid stage", stage)
+	}
+
+	c.t.s.NoError(<-updateErrCh)
+}
+
+func (c *hrsuTestCluster) waitUpdateAdmitted(tv *testvars.TestVars, updateID string) {
+	c.t.s.Eventuallyf(func() bool {
```

Comment:
Better if you use `EventuallyWithT` and put assertions that the `pollErr` is nil. Otherwise you'll be blind to such errors.